zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -0,0 +1,524 @@
+        sum_to_one = sum_to_one + switch.expr();
+    }
+
+    constraints.push(1.expr() - sum_to_one);
+
+    constraints
+}
+
+#[derive(Debug)]
+struct CaseConfig {
+    case: Case,
+    num_word: usize,
+    num_cell: usize,
+    will_halt: bool,
+}
+
+impl CaseConfig {
+    fn num_total_cell(&self) -> usize {
+        32 * self.num_word
+            + self.num_cell
+            + if self.will_halt {
+                NUM_CELL_RESUMPTION
+            } else {
+                0
+            }
+    }
+
+    // allocate indexes of cells for words, cells and unused. It assumes input
+    // free_cells are in order of rotation and then column index.
+    fn allocation_idxs<F: FieldExt>(
+        &self,
+        num_case: usize,
+        free_cells: &[Cell<F>],
+    ) -> (Vec<Range<usize>>, Vec<usize>, Vec<usize>) {
+        let num_free_cell = free_cells.len();
+        let num_total_cell = num_case + self.num_total_cell();
+        assert!(num_free_cell >= num_total_cell);
+
+        let num_unused = num_free_cell - num_total_cell;
+        let word_height = 32 / CIRCUIT_WIDTH;
+
+        let mut word_ranges = Vec::with_capacity(self.num_word);
+        let mut cell_idxs = Vec::with_capacity(self.num_cell + num_unused);
+
+        let mut begin_idx = num_case;
+        let mut rotation_range = Range::default();
+        for (idx, cell) in free_cells.iter().enumerate() {
+            // skip case selector
+            if idx < num_case {
+                continue;
+            }
+
+            let next_idx = idx + 1;
+            if next_idx == num_free_cell || word_ranges.len() == self.num_word {
+                cell_idxs.append(&mut (idx..num_free_cell).collect());
+                break;
+            }
+
+            // find cells in rectangle region for words
+            if rotation_range.contains(&cell.rotation) {
+                if next_idx - begin_idx == 32 {
+                    let next_cell = &free_cells[next_idx];
+                    word_ranges.push(begin_idx..next_idx);
+                    begin_idx = next_idx;
+                    rotation_range =
+                        next_cell.rotation..next_cell.rotation + word_height;
+                }
+            } else {
+                cell_idxs.append(&mut (begin_idx..idx).collect());
+                begin_idx = idx;
+                rotation_range = cell.rotation..cell.rotation + word_height;
+            }
+        }
+
+        assert_eq!(
+            word_ranges.len(),
+            self.num_word,
+            "not enough rectangle regions for words"
+        );
+
+        let (cell_idxs, unused_idxs) = cell_idxs.split_at(self.num_cell);
+        (word_ranges, cell_idxs.to_vec(), unused_idxs.to_vec())
+    }
+}
+
+#[derive(Debug)]
+struct CaseAllocation<F> {
+    selector: Cell<F>,
+    words: Vec<Word<F>>,
+    cells: Vec<Cell<F>>,
+    resumption: Option<Resumption<F>>,
+}
+
+#[derive(Clone, Debug)]
+struct Resumption<F> {
+    caller_id: Cell<F>,
```

Comment:
Use `call_id` to be consistent with `call_id` in the `OpExecutionState`?

```suggestion
    call_id: Cell<F>,
```

Reply:
This is intended for caller's id instead of current call's id, call should resume to caller's execution after halting if it's a internal call. Will add more document on this when I really use it in the cases which halt.

Reply:
I see. Thanks for the clarification.